castervoice/lib/text_manipulation_functions.py

```diff
@@ -0,0 +1,476 @@
+from dragonfly import Key, Pause, AppContext, Window
+import pyperclip
+import re 
+import copy
+from castervoice.lib import context
+from castervoice.lib.ccr.core.punctuation import text_punc_dict,  double_text_punc_dict
+from castervoice.lib.alphanumeric import caster_alphabet
+
+
+new_text_punc_dict = copy.deepcopy(text_punc_dict)
+new_text_punc_dict.update(caster_alphabet)
+character_dict = new_text_punc_dict
+character_list = character_dict.values()
+
+contexts = {
+    "texstudio": AppContext(executable="texstudio"),
+    "lyx": AppContext(executable="lyx")
+}
+
+def get_application():
+    window = Window.get_foreground()
+    # Check all contexts. Return the name of the first one that matches or
+    # "standard" if none matched.
+    for name, context in contexts.items():
+        if context.matches(window.executable, window.title, window.handle):
+            return name
+    return "standard"
+
+def get_start_end_position(text, phrase, direction, occurrence_number):
+# def get_start_end_position(text, phrase, direction):
+    if phrase in character_list:
+        pattern = re.escape(phrase)
+    else:
+        # avoid e.g. matching 'and' in 'land' but allow e.g. matching 'and' in 'hello.and'
+        # for matching purposes use lowercase
+        # PROBLEM: this will not match words in class names like "Class" in "ClassName"
+        # PROBLEM: it's not matching the right one when you have two occurrences of the same word in a row
+        pattern = '(?:[^A-Za-z]|\A)({})(?:[^A-Za-z]|\Z)'.format(phrase.lower()) # must get group 1
+
+    if not re.search(pattern, text.lower()):
+        # replaced phase not found
+        print("'{}' not found".format(phrase))
+        return
+    match_iter = re.finditer(pattern, text.lower())
+    if phrase in character_list: # consider changing this to if len(phrase) == 1 or something
+        match_index_list = [(m.start(), m.end()) for m in match_iter] 
+    else:
+        match_index_list = [(m.start(1), m.end(1)) for m in match_iter] # first group
+    
+    if direction == "left":
+        try:
+            match = match_index_list[-1*occurrence_number] # count from the right
+        except IndexError:
+            print("There aren't that many occurrences of '{}'".format(phrase))
+            return
+    if direction == "right":
+        try:
+            match = match_index_list[occurrence_number - 1] # count from the left
+        except IndexError:
+            print("There aren't that many occurrences of '{}'".format(phrase))
+            return 
+    left_index, right_index = match
+    return (left_index, right_index)
+    
+copy_pause_time_dict = {"standard": "10", "texstudio": "70", "lyx": "60"}
+paste_pause_time_dict = {"standard": "0", "texstudio": "100", "lyx": "20"}
```

Comment:
this doesn't seem to ever get used

Reply:
> this doesn't seem to ever get used

I'm having some error on Github so I can't see the file, but on my version those dictionaries are used in the functions text_manipulation_copy and the text_manipulation_paste.